core/database.py

```diff
@@ -342,18 +345,56 @@ def neighbors(self, neighbor_type=""):
             info[node.full_type] = info.get(node.full_type, []) + [(link, node)]
         return info
 
-    def neighbors_advanced(self, klass, filter, regex, ignorecase, page, rng):
-        from core.web.helpers import get_queryset
-
-        out = [(l, l.dst) for l in Link.objects(__raw__={"src.$id": self.id, "dst.cls": re.compile(klass._class_name)}).no_dereference()]
-        inc = [(l, l.src) for l in Link.objects(__raw__={"dst.$id": self.id, "src.cls": re.compile(klass._class_name)}).no_dereference()]
-
-        all_links = {ref.id: link for link, ref in inc + out}
-        filter['id__in'] = all_links.keys()
-
-        objs = list(get_queryset(klass, filter, regex, ignorecase).limit(rng).skip(page * rng))
-
-        final_list = [(all_links[obj.id], obj) for obj in objs]
+    def _neighbors_aggregation(self, way, klass, result_filters, regex, skip, limit):
+        if way == "in":
+            e1, e2 = "dst", "src"
+        if way == "out":
+            e1, e2 = "src", "dst"
+        collection_name = klass._get_collection().name
+
+        match = {"$match": result_filters}
+
+        pipeline = [
+          {"$match": {"{}.$id".format(e1): self.id, "{}.$ref".format(e2): collection_name}},
+          {"$project": {"_id": True, "src": True, "dst": True, "history": True, "oid": {"$arrayElemAt": [{"$objectToArray": "${}".format(e2)},1]}}},
+          {"$project": {"_id": True, "src": True, "dst": True, "history": True, "oid": "$oid.v"}},
+          {"$lookup": {
+            "from": collection_name,
+            "localField": "oid",
+            "foreignField": "_id",
+            "as": "related",
+          }},
+          {"$unwind" : "$related"},
+        ]
+        pipeline.extend([match, {"$skip": skip*limit}, {"$limit": limit}])
+        results = list(Link.objects.aggregate(*pipeline))
+        return results
+
+    def neighbors_advanced(self, klass, filters, regex, ignorecase, page, rng):
+        result_filters = dict()
+
+        for key, value in filters.items():
+            key = key.replace(".", "__")
+            if key in SEARCH_REPLACE:
+                key = SEARCH_REPLACE[key]
+
+            if regex and isinstance(value, basestring):
+                value = {"$regex": value}
+                if ignorecase:
+                    value["$options"] = "i"
+            if isinstance(value, list) and not key.endswith("__in"):
+                key += "__all"
+            result_filters["related."+key] = value
```

Comment:
🚎  When you copy/paste code from somewhere else, this ends up being duplicated code 🚎

Reply:
More importantly, the `__` prefix is invalid when throwing things into the aggregation pipeline. Will fix.

Reply:
The code changes significantly when hitting the mongodb aggregation pipeline directly vs. crafting a Mongoengine queryset